fairscale/optim/layerwise_gradient_scaler.py

```diff
@@ -0,0 +1,278 @@
+        model,
+        layer_scale_dict: dict,
+        growth_factor: float = 2.0,
+        backoff_factor: float = 0.5,
+        growth_interval: int = 10000,
+        min_scaling_factor: float = torch.finfo(torch.float32).tiny,  # type: ignore
+        max_scaling_factor: float = torch.finfo(torch.float32).max,  # type: ignore
+    ) -> None:
+        self._model = model
+        self._layer_scale_dict: dict = layer_scale_dict
+        self._growth_factor: float = growth_factor
+        self._backoff_factor: float = backoff_factor
+        self._growth_interval: int = growth_interval
+        self._apply_layerwise_scaling: bool = True if len(layer_scale_dict.keys()) > 0 else False
+        self._min_scaling_factor = min_scaling_factor
+        self._max_scaling_factor = max_scaling_factor
+        self._handles: List = []
+        self.layer_info: List = []
+
+        if self._apply_layerwise_scaling:
+            assert self._growth_factor > 1.0, "The growth factor must be > 1.0."
+            assert self._backoff_factor < 1.0, "The backoff factor must be < 1.0."
+            self.layer_info = self._build_layer_info()
+
+    def _build_layer_info(self) -> List:
+        """
+        Helper function to create a list of LayerInfo instances.
+        """
+        layer_info_list = list()
+
+        for name, layer in self._model.named_modules():
+            if name != "":
+                if name not in self._layer_scale_dict.keys():
+                    logging.debug("name = %s, layer = %s, scaling_factor = %s" % (name, layer, 1.0))
+                    layer_info_list.append(LayerInfo(name, layer, 1.0))
+                else:
+                    logging.debug(
+                        "name = %s, layer = %s, scaling_factor = %s" % (name, layer, self._layer_scale_dict[name])
+                    )
+                    layer_info_list.append(LayerInfo(name, layer, self._layer_scale_dict[name], True))
+        return layer_info_list
+
+    def scale(self) -> None:
+        """
+        For each layer calculates the scaling factor for preceding layers' grad inputs
+        and current layers' grad outputs. These values are used to register a full backward
+        hook. The handle returned from registering the backward hook is appended to a list
+        of handles. New hooks are created and registered at every step and a new list of
+        handles is created. The handles are flushed out in the unscale function.
+        """
+        if self._apply_layerwise_scaling:
+            for idx in range(len(self.layer_info)):
+                elt = self.layer_info[idx]
+                layer_name, layer_type = elt.layer_name, elt.layer_type
+
+                inputs_multiplier = 1.0
+                if idx > 0:
+                    inputs_multiplier = self.layer_info[idx - 1].scaling_factor
+
+                outputs_multiplier = 1.0 / elt.scaling_factor
+                helper = GradientHelper(layer_name, inputs_multiplier, outputs_multiplier)
```

Comment:
can you explain the inputs_multiplier and outputs_multiplier a bit? what are example values and how do they help scaling?

Reply:
Consider this example graph:

|layer | scaling_factor	|
|--- | --- |
| input	| None 		|
| fc1   	| 128		|
| relu1	| 1 			|
| fc2 	| 512		|
| relu2 	| 1 			|
| sigmoid	| 1 			|
| output |  None |

Following the vocabulary of [register_full_backward_hook](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.register_full_backward_hook),

 **inputs_multiplier = grad_inputs** 
**outputs_multiplier = grad_outputs**.

During the backward pass, the grad_inputs will be multiplied by the scaling factor of the next layer and grad_outputs will be multiplied by the reciprocal of the scaling factor of the current layer.

| layer | inputs_multiplier | outputs_multiplier |
| ---- | --- | --- |
| sigmoid  | 1    | 1     |
|relu2        | 512 | 1 |
| fc2 |     1 | 1/512 |
| relu1 |  128 | 1 |
| fc1 | 1 | 1/128 | 

The inputs_multiplier value is applied to the next layer (going top to bottom in the table above). Note that 

```inputs_multiplier[i] * outputs_multiplier[i+1] = 1```

 This is the scaling and unscaling process.

Note: value of the inputs multiplier in the fc1 layer does not make any difference as there is no layer on which it is applied. 

Edge Case: What if the last layer of a graph has a scaling factor greater than 1? The solution is - append the identity layer to the graph and now we are back in the case described above.

Reply:
Can you add comments and document this somewhere?